crates/phactory/src/prpc_service.rs

```diff
@@ -816,4 +825,273 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> PhactoryApi
+
+        // Share the key with attestation
+        let ecdh_pubkey = challenge_client
+            .decode_ecdh_public_key()
+            .map_err(from_display)?;
+        let iv = crate::generate_random_iv();
+        let (ecdh_pubkey, encrypted_key) = key_share::encrypt_secret_to(
+            &my_identity_key,
+            &[b"worker_key_handover"],
+            &ecdh_pubkey.0,
+            &my_identity_key.dump_secret_key(),
+            &iv,
+        )
+        .map_err(from_debug)?;
+        let encrypted_key = EncryptedKey {
+            ecdh_pubkey: sr25519::Public(ecdh_pubkey),
+            encrypted_key,
+            iv,
+        };
+        let runtime_state = phactory.runtime_state()?;
+        let genesis_block_hash = runtime_state.genesis_block_hash;
+        let payload =
+            pb::EncryptedWorkerKey::new(genesis_block_hash, dev_mode, Some(encrypted_key));
+        let payload_hash = sp_core::hashing::blake2_256(&payload.encode());
+
+        let mut attestation = None;
+        if !dev_mode {
+            let (attn_report, sig, cert) =
+                match phactory.platform.create_attestation_report(&payload_hash) {
+                    Ok(r) => r,
+                    Err(e) => {
+                        let message = format!("Failed to create attestation report: {:?}", e);
+                        error!("{}", message);
+                        return Err(from_display(message));
+                    }
+                };
+            attestation = Some(pb::Attestation {
+                version: 1,
+                provider: "SGX".to_string(),
+                payload: Some(pb::AttestationReport {
+                    report: attn_report,
+                    signature: base64::decode(sig).map_err(from_display)?,
+                    signing_cert: base64::decode_config(cert, base64::STANDARD)
+                        .map_err(from_display)?,
+                }),
+                timestamp: now(),
+            });
+        } else {
+            info!("Omit RA report in workerkey response in dev mode");
+        }
+
+        Ok(pb::GetWorkerKeyResponse {
+            payload: Some(payload),
+            attestation,
+        })
+    }
+
+    // WorkerKey Handover Client
+
+    fn handle_worker_key_challenge(
+        &mut self,
+        request: pb::WorkerKeyChallenge,
+    ) -> RpcResult<pb::WorkerKeyChallengeResponse> {
+        let mut phactory = self.lock_phactory();
+
+        // generate and save tmp key only for key handover encryption
+        let tmp_key = crate::new_sr25519_key();
+        let tmp_ecdh_key = tmp_key
+            .derive_ecdh_key()
+            .expect("should never fail with valid key; qed");
+        let ecdh_public_key = phala_types::EcdhPublicKey(tmp_ecdh_key.public());
+        phactory.tmp_ecdh_key = Some(tmp_ecdh_key);
+
+        let challenge = request.decode_challenge().map_err(from_display)?;
+        let payload = pb::ChallengeClient::new(challenge.clone(), ecdh_public_key);
```

Comment:
`ChallengeClient` the name is not easy to understand

Reply:
Changed to `ChallengeHandler`

Reply:
`ChallengeHandler` sounds like a function to handle the challenge. However, here, I think it refers to the response data to the challenge. Do you think `ChallengeHandlerInfo` makes more sense?

Reply:
Ok, renamed.